### DISCUSSION TITLE
mysql-auth doesn't ban to attacker

### DISCUSSION THREAD

**Discussion by jaimerletona** [fail2ban/fail2ban#3225] [Q&A] [answered]
Hello everyone.
I am using fail2ban 0.11.1-1 on ubuntu 20.04

I confured /etc/fail2ban/jail.local for ssh and other services like mysq-auth. ssh attacker have been blocked like expected. Mysql attackers are don't been banned although the log and fail2ban-regex matches exactly the attack log and type.

**# systemctl status fail2ban.service**
● fail2ban.service - Fail2Ban Service
     Loaded: loaded (/lib/systemd/system/fail2ban.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2022-02-18 14:46:16 CST; 3s ago
       Docs: man:fail2ban(1)
    Process: 27402 ExecStartPre=/bin/mkdir -p /run/fail2ban (code=exited, status=0/SUCCESS)
   Main PID: 27403 (f2b/server)
      Tasks: 17 (limit: 4915)
     Memory: 17.7M
     CGroup: /system.slice/fail2ban.service
             └─27403 /usr/bin/python3 /usr/bin/fail2ban-server -xf start

feb 18 14:46:16 DB_S systemd[1]: Starting Fail2Ban Service...
feb 18 14:46:16 DB_S systemd[1]: Started Fail2Ban Service.
feb 18 14:46:17 DB_S fail2ban-server[27403]: Server ready

**/var/log/mysql/error.log**
2022-02-18T19:15:17.102120Z 164639 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: YES)
2022-02-18T19:40:27.174746Z 169562 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: YES)
2022-02-18T19:44:25.781876Z 169955 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: NO)
2022-02-18T19:44:28.664117Z 169960 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: YES)
2022-02-18T19:44:33.066639Z 169967 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: YES)
2022-02-18T19:44:35.882599Z 169972 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: YES)
2022-02-18T19:44:38.555488Z 169980 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: YES)

**(I was forced to configure mysql to log_error_verbosity = 3 to see the error log)**

Extract of **/etc/fail2ban/jail.local**
[mysqld-auth]
enabled = true
port     = 3306
#logpath  = %(mysql_log)s
logpath  = /var/log/mysql/error.log
backend  = %(mysql_backend)s
maxretry = 3
findtime = 5d
bantime = 1d
filter =  mysqld-auth

I did try test the regexp and matches, but not banned anything
Result of fail2banregex command

**# fail2ban-regex --print-all-matches /var/log/mysql/error.log /etc/fail2ban/filter.d/mysqld-auth.conf**
Usage: /usr/bin/fail2ban-regex [OPTIONS] <LOG> <REGEX> [IGNOREREGEX]

fail2ban-regex: error: no such option: --print-all-matches
root@DB_S /etc/fail2ban# fail2ban-regex --print-all-matched /var/log/mysql/error.log /etc/fail2ban/filter.d/mysqld-auth.conf 

Running tests

Use   failregex filter file : mysqld-auth, basedir: /etc/fail2ban
Use      datepattern : Default Detectors
Use         log file : /var/log/mysql/error.log
Use         encoding : UTF-8

Results

Failregex: 43 total
|-  #) [# of hits] regular expression
|   1) [43] ^(?:\[\])?\s*(?:<[^.]+\.[^.]+>\s+)?(?:\S+\s+)?(?:kernel:\s?\[ *\d+\.\d+\]:?\s+)?(?:@vserver_\S+\s+)?(?:(?:(?:\[\d+\])?:\s+[\[\(]?\S*(?:\(\S+\))?[\]\)]?:?|[\[\(]?\S*(?:\(\S+\))?[\]\)]?:?(?:\[\d+\])?:?)\s+)?(?:\[ID \d+ \S+\]\s+)?(?:(?:\d{6}|\d{4}-\d{2}-\d{2})[ T]\s?\d{1,2}:\d{2}:\d{2} )?(?:\d+ )?\[\w+\] (?:\[[^\]]+\] )*Access denied for user '[^']+'@'<HOST>' (to database '[^']*'|\(using password: (YES|NO)\))*\s*$
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [52] {^LN-BEG}ExYear(?P<_sep>[-/.])Month(?P=_sep)Day(?:T|  ?)24hour:Minute:Second(?:[.,]Microseconds)?(?:\s*Zone offset)?
`-

Lines: 52 lines, 0 ignored, 43 matched, 9 missed
[processed in 0.00 sec]

|- Matched line(s):
password: YES)
|  2022-02-18T19:44:57.789593Z 170012 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: YES)
|  2022-02-18T19:45:01.150181Z 170015 [Note] [MY-010926] [Server] Access denied for user 'TTT'@'10.120.120.15' (using password: YES)
...
|- Missed line(s):
|  2022-02-18T18:39:46.015766Z 157239 [Warning] [MY-010055] [Server] IP address '10.120.120.15' could not be resolved: Name or service not known
|  2022-02-18T18:39:46.991594Z 157239 [Warning] [MY-013360] [Server] Plugin sha256_password reported: ''sha256_password' is deprecated and will be removed in a future release. Please use caching_sha2_password instead'
...

So, I see that everything is OK, but it doesn't work.
How can I trace or test if fail2ban is triggering the process when the log apply to the filter?
In advance, I'm thankful for any orientation.

**Comment by sebres** [comment, marked as answer]
What do you see in fail2ban.log for `mysqld-auth`?

> `backend = %(mysql_backend)s`

this is set by your fail2ban distribution and could be wrong if it is set to systemd (since you need to monitor log-file and not a journal).
What shows `fail2ban-client -d | grep "'add'.*mysqld-auth"` for you?
If it is `['add', 'mysqld-auth', 'systemd']` then it is not correct.
Set `backend = auto` there (it would use file-related backends instead).

> How can I trace or test if fail2ban is triggering the process when the log apply to the filter?

In fail2ban.log you'd see `[mysqld-auth] Found 192.0.2.1` for every finding of IP 192.0.2.1 and `[mysqld-auth] Ban 192.0.2.1` for a ban.
See https://github.com/fail2ban/fail2ban/wiki/How-fail2ban-works for details.

If you mean you may need more verbose logging you can try this:
```
# set runtime log-level to DEBUG or HEAVYDEBUG
?sudo? fail2ban-client set loglevel DEBUG
?sudo? fail2ban-client set loglevel HEAVYDEBUG
```
don't forget to restore log-level back to INFO hereafter:
```
?sudo? fail2ban-client set loglevel INFO
```

And last but not least - `findtime = 5d` is too heavy (do you really need to consider 3 attempts in 5 days?)
If you hope to find some old failures after restart, it would not work in fail2ban this way - it would ignore already processed entries (fail2ban stores last known position in log and seeks to it after restart) and it would ignore all entries older than 1 day (due to `bantime = 1d`, because the ban would be outdated).

**Reply by jaimerletona** [reply]
Thanks for your help, in fact I had configured wrong the backend = %(mysql_backend)s 

Before:
fail2ban-client -d | grep "'add'.*mysqld-auth"
['add', 'mysqld-auth', 'systemd']

After:
fail2ban-client -d | grep "'add'.*mysqld-auth"
['add', 'mysqld-auth', 'auto']

I corrected that, and the findtime, and now works ok.
THANKS A LOT!!!

**Reply by fyt0907** [reply]
My situation is exactly the same as his, but after I set backend=auto, fail 2ban doesn't work anymore Can you help me see what to do?

jail.local
![jl](https://github.com/user-attachments/assets/05bedd19-0b70-4c76-aabe-fef0431731bf)
fail2ban.log
![fail2ban log](https://github.com/user-attachments/assets/da2a6ba3-ccf2-4526-8943-674fc8d774bb)
fail2ban service status
![st](https://github.com/user-attachments/assets/0343736b-6c57-4d20-9ca8-58845c32d3ff)

**Reply by sebres** [reply]
The error message `Failed during configuration: Have not found any log file for mysqld-auth jail` is pretty clear - fail2ban cannot find the logs to monitor. Looks like `logpath` value is incorrect and doesn't match any file - simply set it to the fullpath, where your mysqld is currently logging.
And last but not least, it is much easier to analyze a copy&pasted texts (configs/errors/etc) instead of the screenshot pictures, so please avoid it henceforth if not really necessary.

**Reply by fyt0907** [reply]
It works!
Thanks!

**Comment by SuoodM** [comment]
Am Having a similar issue,

am trying to configure mysql in fail2ban, and I can see the fail attempts that am testing out from the log that I have created, but the issue is fail2ban is not banning the attacker and nothing shows on the failing attempts when I run "fail2ban-client status mysqld-auth". 

I have tried the backend auto just in case but it doesn't work, btw ssh configuration work just fine.

Here is my configuration:

This from jail.conf:
![Screenshot 2025-06-29 013219](https://github.com/user-attachments/assets/eb7a99e2-3a15-40d9-a6e6-1f25253e83cc)
mysqld configuration
![image](https://github.com/user-attachments/assets/e8827c22-a885-4b66-99cf-8aa25073ca5d)
fileregex just in case
![image](https://github.com/user-attachments/assets/44249d12-c7d7-45bf-b68b-2cf6f45d950d)
Here are the failing attempts from mysql.log
![image](https://github.com/user-attachments/assets/515a7902-59bb-4119-9f5a-5d50bd6189ca)

![image](https://github.com/user-attachments/assets/bacb75df-be01-437a-80a4-ed1c7118f6c3)
After the multiple failed attempts nothing happened
![image](https://github.com/user-attachments/assets/6dc51bc5-0702-48d2-9c96-8d26f7d114a8)

**Reply by sebres** [reply]
Firstly, please don't use the screenshots for plain text (use copy&paste of text instead) - the pictures make the analyze more difficult for us, e. g. more work (typing by hand, etc).

The logging format is almost completely different as covered by filter - it doesn't have prefix at all (host, pid, etc) and has padding by tokens, tokens not enclosed in square brackets, etc. Thus current filter is not suitable for this format at all.

You can use something like this (instead of filter):
```ini
[mysqld-auth]
filter = 
failregex = ^\s*\d+\s+Connect\s+Access denied for user '<F-USER>[^']+</F-USER>'@'<ADDR>'
...
```
